File: ibkr_show_frontend/src/types/symbolAnalysis.ts
export interface FinancialField {
  id: string
  name: string
  value: number | string | null
  raw_value: string | null
  yoy: number | null
  value_type: string | null
  level: number | null
}

export interface FinancialPeriod {
  label: string
  fiscal_year: number | null
  fiscal_period: string
  report_type: string
  metrics: Record<string, number | null>
  statements: Record<string, FinancialField[]>
}

export interface SymbolMarketSnapshot {
  symbol: string
  name: string | null
  currency: string | null
  last_price: number | null
  change_percent: number | null
  market_cap: number | null
  pe_ttm: number | null
  forward_pe: number | null
  pe_3y_median: number | null
  pe_industry_median: number | null
  pb: number | null
  dividend_yield: number | null
  turnover_rate: number | null
  eps_ttm: number | null
  forward_eps: number | null
  bps: number | null
  total_shares: number | null
  valuation_date: string | null
  valuation_summary: string | null
}

export interface SymbolFinancialsResponse {
  symbol: string
  currency: string | null
  report_type: string
  period_count: number
  periods: FinancialPeriod[]
  market_snapshot: SymbolMarketSnapshot | null
  source: string
}

export interface MetricComparisonItem {
  key: string
  label: string
  left_value: number | null
  right_value: number | null
  winner: 'left' | 'right' | 'tie' | 'unknown' | string
}

export interface SymbolComparisonResponse {
  left: SymbolFinancialsResponse
  right: SymbolFinancialsResponse
  latest_metric_comparison: MetricComparisonItem[]
}

export interface SymbolAiAdviceResponse {
  left_symbol: string
  right_symbol: string
  recommendation: 'left' | 'right' | 'neutral' | string
  confidence: 'high' | 'medium' | 'low' | string
  summary: string
  key_reasons: string[]
  risks: string[]
  add_conditions: string[]
  data_limitations: string[]
  raw_response: string | null
}

File: ibkr_show_backend/app/schemas/symbol_analysis.py
from pydantic import BaseModel, Field


class FinancialField(BaseModel):
    id: str
    name: str
    value: float | str | None = None
    raw_value: str | None = None
    yoy: float | None = None
    value_type: str | None = None
    level: int | None = None


class FinancialPeriod(BaseModel):
    label: str
    fiscal_year: int | None = None
    fiscal_period: str
    report_type: str
    metrics: dict[str, float | None] = Field(default_factory=dict)
    statements: dict[str, list[FinancialField]] = Field(default_factory=dict)


class SymbolMarketSnapshot(BaseModel):
    symbol: str
    name: str | None = None
    currency: str | None = None
    last_price: float | None = None
    change_percent: float | None = None
    market_cap: float | None = None
    pe_ttm: float | None = None
    forward_pe: float | None = None
    pe_3y_median: float | None = None
    pe_industry_median: float | None = None
    pb: float | None = None
    dividend_yield: float | None = None
    turnover_rate: float | None = None
    eps_ttm: float | None = None
    forward_eps: float | None = None
    bps: float | None = None
    total_shares: float | None = None
    valuation_date: str | None = None
    valuation_summary: str | None = None


class SymbolFinancialsResponse(BaseModel):
    symbol: str
    currency: str | None = None
    report_type: str
    period_count: int
    periods: list[FinancialPeriod]
    market_snapshot: SymbolMarketSnapshot | None = None
    source: str = "longbridge"


class MetricComparisonItem(BaseModel):
    key: str
    label: str
    left_value: float | None = None
    right_value: float | None = None
    winner: str


class SymbolComparisonResponse(BaseModel):
    left: SymbolFinancialsResponse
    right: SymbolFinancialsResponse
    latest_metric_comparison: list[MetricComparisonItem]


class SymbolAiAdviceRequest(BaseModel):
    left_symbol: str
    right_symbol: str
    question: str | None = None


class SymbolAiAdviceResponse(BaseModel):
    left_symbol: str
    right_symbol: str
    recommendation: str
    confidence: str
    summary: str
    key_reasons: list[str] = Field(default_factory=list)
    risks: list[str] = Field(default_factory=list)
    add_conditions: list[str] = Field(default_factory=list)
    data_limitations: list[str] = Field(default_factory=list)
    raw_response: str | None = None
